API/Controllers/TournamentsController.cs

```diff
@@ -20,68 +22,89 @@ IAdminNoteService adminNoteService
         if (await tournamentsService.ExistsAsync(tournamentSubmission.Name, tournamentSubmission.Ruleset))
         {
-            return BadRequest(
-                $"A tournament with name {tournamentSubmission.Name} for ruleset {tournamentSubmission.Ruleset} already exists"
+            ModelState.AddModelError(
+                nameof(TournamentSubmissionDTO.Name),
+                $"A tournament with the name '{tournamentSubmission.Name}' and ruleset " +
+                $"'{tournamentSubmission.Ruleset}' already exists."
+            );
+            ModelState.AddModelError(
+                nameof(TournamentSubmissionDTO.Ruleset),
+                $"A tournament with the ruleset '{tournamentSubmission.Ruleset}' and name " +
+                $"'{tournamentSubmission.Name}' already exists."
             );
         }
 
-        if (tournamentSubmission.RejectionReason.HasValue && !User.IsAdmin())
+        // Some submission data requires admin permission
+        if (!User.IsAdmin())
         {
-            return BadRequest("Only admin users may supply a rejection reason");
+            // Check reject on submit
+            if (tournamentSubmission.RejectionReason.HasValue)
+            {
+                ModelState.AddModelError(
+                    nameof(TournamentSubmissionDTO.RejectionReason),
+                    $"The field {nameof(TournamentSubmissionDTO.RejectionReason)} may only be supplied by admin users."
+                );
+            }
+
+            // We never want unprivileged users submitting empty tournaments
+            if (!tournamentSubmission.Ids.Any())
+            {
+                ModelState.AddModelError(
+                    nameof(TournamentSubmissionDTO.Ids),
+                    $"The field {nameof(TournamentSubmissionDTO.Ids)} cannot be empty."
+                );
+            }
         }
 
-        // Remove query string
-        try
+        // Check for valid forum post url
+        if (!Uri.TryCreate(tournamentSubmission.ForumUrl, UriKind.Absolute, out Uri? sanitizedUri))
         {
-            tournamentSubmission.ForumUrl = new Uri(tournamentSubmission.ForumUrl).GetLeftPart(UriPartial.Path);
+            ModelState.AddModelError(
+                nameof(TournamentSubmissionDTO.ForumUrl),
+                $"The field {nameof(TournamentSubmissionDTO.ForumUrl)} is invalid."
+            );
         }
-        catch (UriFormatException)
+
+        if (!ModelState.IsValid)
         {
-            return BadRequest($"The tournament forum URL is invalid: {tournamentSubmission.ForumUrl}");
+            return ValidationProblem();
         }
 
-        // Input validation for mp and beatmap ids
-        tournamentSubmission.Ids = tournamentSubmission.Ids.Where(id => id >= 1);
-        tournamentSubmission.BeatmapIds = tournamentSubmission.BeatmapIds.Where(id => id >= 1);
+        #endregion
+
+        // Sanitize forum url
+        tournamentSubmission.ForumUrl = sanitizedUri!.AbsolutePath;
```

Comment:
Why does `Uri.TryCreate` produce a nullable `Uri` at all? I feel like it shouldn't be assumed the Uri is non-null here if the `Uri.TryCreate` method produces a nullable `out`.

Reply:
This is the definition for the method

```cs
public static bool TryCreate([NotNullWhen(true), StringSyntax(StringSyntaxAttribute.Uri, "uriKind")] string? uriString, UriKind uriKind, [NotNullWhen(true)] out Uri? result)
```

99% of the time the case with `Try{X}` methods that return a `bool` and produce an `out` parameter is that the `out` parameter _has to be nullable_. You can work with this and improve code quality by using a `System.Diagnostics.CodeAnalysis.NotNullWhenAttribute` which can flag to analyzers that the `out` value will always be present based on the return value of the method.

Since we do this:

```cs
        // Check for valid forum post url
        if (!Uri.TryCreate(tournamentSubmission.ForumUrl, UriKind.Absolute, out Uri? sanitizedUri))
        {
            ModelState.AddModelError(
                nameof(TournamentSubmissionDTO.ForumUrl),
                $"The field {nameof(TournamentSubmissionDTO.ForumUrl)} is invalid."
            );
        }

        if (!ModelState.IsValid)
        {
            return ValidationProblem();
        }

        // Sanitize forum url
        tournamentSubmission.ForumUrl = sanitizedUri!.AbsolutePath;
```
It is not clear to the analyzer whether or not the result of the method caused a return or branching flow. However we do know for sure that if the result is false, the statement `tournamentSubmission.ForumUrl = sanitizedUri!.AbsolutePath` would never execute because of the check to the model state.

Reply:
I actually implemented a method with the same design as what I am describing in my last batch of changes
```cs
    public static bool TryGetSubjectId(
        this ClaimsPrincipal claimsPrincipal,
        [NotNullWhen(true)] out int? subjectId
    )
    {
        subjectId = null;
        try
        {
            subjectId = claimsPrincipal.GetSubjectId();
            return true;
        }
        catch (ArgumentException)
        {
            return false;
        }
    }
```

Reply:
Wow, very interesting. Thanks for the explanation.